src/main/java/org/primefaces/component/datatable/DataTableRenderer.java

```diff
@@ -889,14 +889,20 @@ protected void encodeDefaultFilter(FacesContext context, DataTable table, UIColu
     protected void encodeFilterInput(UIColumn column, ResponseWriter writer, boolean disableTabbing,
         String filterId, String filterStyleClass, Object filterValue, String ariaLabelId) throws IOException {
 
-        if (column.getValueExpression(Column.PropertyKeys.filterOptions.toString()) == null) {
+        if (hasFilterOptions(column)) {
             encodeFilterInputText(column, writer, disableTabbing, filterId, filterStyleClass, filterValue, ariaLabelId);
         }
         else {
             encodeFilterInputSelect(column, writer, disableTabbing, filterId, filterStyleClass, filterValue, ariaLabelId);
         }
     }
 
+    private boolean hasFilterOptions(UIColumn column) {
+        return column.getValueExpression(Column.PropertyKeys.filterOptions.toString()) == null
+            || getFilterOptions(column) == null
+            || getFilterOptions(column).length == 0;
+    }
```

Comment:
My only comment is you could assign `getFilterOptions(column)` in a variable and use that variable twice.

Reply:
Your're right. And in addition, the method name `hasFilterOptions()` was misleading as it was actually checking if no filterOptions were present (logic was correct, though). I'll update the PR.

Reply:
Done, PR has been updated.